zapcore/encoder.go

```diff
@@ -170,6 +181,8 @@ func (e *TimeEncoder) UnmarshalText(text []byte) error {
 		*e = EpochMillisTimeEncoder
 	case "nanos":
 		*e = EpochNanosTimeEncoder
+	case "format=":
+		*e = TimeEncoderOfFormat(vals[1])
```

Comment:
I see the concern about the `=` splitting thing. This could become a problem
if we want more parameters in the future, or if escaping parameters becomes
necessary.

Perhaps we can accept objects for when the format is specified:

```
timeEncoder: rfc3339
# or
timeEncoder:
  format: "06/01/02 03:04pm"
```

We can accomplish this by implementing [`yaml.Unmarshaler`][2] to decode into
an object, and if that fails, to fall back to UnmarshalText.

  [1]: https://golang.org/pkg/encoding/json/#Unmarshaler
  [2]: https://godoc.org/gopkg.in/yaml.v2#Unmarshaler

Something like the following should suffice:

    func (e *TimeEncoder) UnmarshalYAML(unmarshal func(interface{}) error) error {
      # timeEncoder:
      #   format: "..."
      var o struct {
        Format string `yaml:"format" json:"format"`
      }
      if err := unmarshal(&o); err == nil {
        *e = TimeEncoderOfFormat(o.Format)
        return nil
      }
      
      # timeEncoder: "..."
      var s string
      if err := unmarshal(&s); err != nil {
        return err
      }
      return e.UnmarshalText([]byte(s))
    }

We can even reuse the `UnmarshalYAML` method for JSON support:

    func (e *TimeEncoder) UnmarshalJSON(bs []byte) error {
      return e.UnmarshalYAML(func(v interface{}) error {
        return json.Unmarshal(bs, v)
      })
    }

Reply:
Seems more expressive!
Fixed at 0c345e2

Also, modified tests so that it explicitly uses yaml.Unmarshal (5f2bb49)